### PR TITLE
fix(learn): advanced-node-and-express username field

### PR DESCRIPTION
**Pull request by Xavier-Pierre-dev**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

I notice that for the moment the curriculum have a little coherency issue starting at github authentication. In reality at the start of the project we set up our strategy with the `username` property for example when we register an user using the form the curriculum will push us to use : 
```js
app.route('/register')
  .post((req, res, next) => {
    myDataBase.findOne({ username: req.body.username }, function(err, user) {
      if (err) {
        next(err);
      } else if (user) {
        res.redirect('/');
      } else {
        myDataBase.insertOne({
          username: req.body.username,
          password: req.body.password
        },
          (err, doc) => {
            if (err) {
              res.redirect('/');
            } else {
              // The inserted document is held within
              // the ops property of the doc
              next(null, doc.ops[0]);
            }
          }
        )
      }
    })
  },
    passport.authenticate('local', { failureRedirect: '/' }),
    (req, res, next) => {
      res.redirect('/profile');
    }
  );
```

As you can see we use `username: req.body.username` then we also use that to display content on profile.pug. In reality `username` property are use since start until the github authentication. In fact for github authentication `username` property are not set up so later with socket we use `name` instead of `username` this lead to an issue where if we login with something else than github the name will be undefined for the chat app and not be displayed. 

This is why I propose some change inside the curriculum in order to add the `username` property when a user use github in order to register himself and edit the next step so the challenge guideline will push the learner to use the username instead of the user. To make it properly work either it's with simple register form, login with or without github

### CODE REVIEW

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/send-and-display-chat-messages.md
Comment (hanswang123456) [inline curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/send-and-display-chat-messages.md:25]: Id change the last sentence to:
Once the event is received, it should emit the event `'chat message'` to all sockets using `io.emit`, sending a data object containing the `username` and `message`.

Comment (Xavier-Pierre-dev) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/send-and-display-chat-messages.md:25]: I think your sentence is shorter and easier to read. Should I add this change too?

**Comment by Xavier-Pierre-dev** [comment]
Sorry I make a mistake with that : 
[Xavier-Pierre-dev](https://github.com/Xavier-Pierre-dev) requested review from [hanswang123456](https://github.com/hanswang123456) and removed request for [ShaunSHamilton](https://github.com/ShaunSHamilton) [8 minutes ago](https://github.com/freeCodeCamp/freeCodeCamp/pull/47349#event-7244232028)

It wasn't intentional, I clicked by mistake, this is my first PR.

**Review by hanswang123456 — COMMENTED** [review]
Looks Good!

**Comment by hanswang123456** [comment]
@Xavier-Pierre-dev  Please get @ShaunSHamilton, @Sboonny, or @moT01 to review as they are actual reviewers. I'm just looking out for any small fixes.

**Comment by hanswang123456** [comment]
Also check out issue https://github.com/freeCodeCamp/freeCodeCamp/issues/47109. Might be a nice issue for you to attempt.

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Hello there,

I am requesting changes to this PR to block it for now, because this will require the relevant Gists to be updated as well - something a member of staff will have to do.

**Comment by raisedadead** [comment]
@ShaunSHamilton let's review this and do any additional tasks. This has been sitting here for a while.

**Comment by ShaunSHamilton** [comment]
To mention, this is related to: https://github.com/freeCodeCamp/freeCodeCamp/issues/39599

**Comment by ShaunSHamilton** [comment]
This is ready to go in. Once we are production ready, we need to have the Gists updated based of of mine here: https://gist.github.com/ShaunSHamilton

Also, the boilerplate needs updating: <insert_pr_once_made>

**Comment by raisedadead** [comment]
> Once we are production ready

Do you want to target Monday afternoon your time?

**Comment by ShaunSHamilton** [comment]
> Do you want to target Monday afternoon your time?

Yes, that suits me.

**Comment by raisedadead** [comment]
It's a date. You are paying for my beer.

**Comment by ieahleen** [comment]
> Do you want to target Monday afternoon your time?

Was that yesterday?

**Comment by raisedadead** [comment]
> Was that yesterday?

Yes - we met, but we could not get this in. Also, we decided that this needs review from others. We will talk about this in the dev-team catch-up meeting.

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/announce-new-users.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/announce-new-users.md:13]: I think this might be a typo. It seems like the text here and code below are referring to the `io.emit` event's name, which should be change to `'user'` in this step:

```suggestion
Change the event name to `'user'`, and pass an object along containing the fields `username`, `currentUsers`, and `connected` (to be `true` in case of connection, or `false` for disconnection of the user sent). Be sure to change both `'user count'` events and set the disconnect one to send `false` for the field `connected` instead of `true` like the event emitted on connect.
```

**Review by moT01 — CHANGES_REQUESTED** [review]
This all looks real good @Xavier-Pierre-dev 🎉

I went through the whole thing pretty thoroughly. The project is pretty awesome - I never actually went through it like that.

I made quite a few suggestions to try and improve the instructions. Some of them were typos or other grammar suggestions in your additions, but most were because I was having a hard time differentiating explanation from instructions of what I was supposed to do. So it's mostly an attempt to clarify to campers what they are actually supposed to do or add - which I think is overdue. Also, the instructions for creating an app on GH needed some updating. Many of these were pre-existing and possibly outside the scope of this PR, but I think we should add them here. Let me know if you have any questions on any of it.

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/serialization-of-a-user-object.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/serialization-of-a-user-object.md:22]: Oops 😆

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/serialization-of-a-user-object.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/serialization-of-a-user-object.md:26]: I do this typo far too often.

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/use-a-template-engines-powers.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/use-a-template-engines-powers.md:15]: All of these changes were out of the scope of this PR. It just felt like the easiest time to make them 🤷‍♂️

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/registration-of-new-users.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/registration-of-new-users.md:65]:
```suggestion
You should have a `/register` route and display a registration form on the home page.
```

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/registration-of-new-users.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/registration-of-new-users.md:80]:
```suggestion
    'You should have a route that accepts a POST request on /register that queries the db with findOne and the query being username: req.body.username'
```

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/hashing-your-passwords.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/hashing-your-passwords.md:17]: Only if the earlier suggestion for simplifying the checks and early returns in `passport.use(new LocalStrategy...` is accepted:

```suggestion
On your authentication strategy, you check for the following in your code before completing the process: `if (password !== user.password) return done(null, false);`. After making the previous changes, now `user.password` is a hash. Before making a change to the existing code, notice how the statement is checking if the password is **not** equal then return non-authenticated. With this in mind, change that code to look as follows to properly check the password entered against the hash:
```

**Review by scissorsneedfoodtoo — COMMENTED** [review]
I was only able to get through about half of the project, but everything is looking much better than before.

A lot of the things I noticed are pretty minor, so I'd be fine with disregarding a lot of them. Just wanted to point them out since there are a lot of other improvements here.

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/use-a-template-engines-powers.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/use-a-template-engines-powers.md:45]: Sigh... now I need to edit the Gists 😒

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/authentication-strategies.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/authentication-strategies.md:30]: This was just a personal preference.

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/implement-the-serialization-of-a-passport-user.md
Comment (ShaunSHamilton) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/implement-the-serialization-of-a-passport-user.md:21]: Something else below. It made sense when it was first introduced. We could completely remove the comments - they are only there because, unless a Camper removes them, they will likely have them in their code from earlier.

Comment (ShaunSHamilton) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/implement-the-serialization-of-a-passport-user.md:21]: I removed these from the Gists. So, we can remove them here too.

**Review by moT01 — DISMISSED** [review]
LGTM @ShaunSHamilton 🎉

File: curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/implement-the-serialization-of-a-passport-user.md
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/implement-the-serialization-of-a-passport-user.md:21]: Awesome. Thanks for confirming that, and for going through all the effort to update those Gists.

I think the comments can be useful when this block of code is first introduced in the lessons, but it would be easier to remove them all from the Gists.

**Review by scissorsneedfoodtoo — DISMISSED** [review]
Thanks for all of your hard work on this @Xavier-Pierre-dev, and all the other improvements @ShaunSHamilton and @moT01 made look great. The instructions are much easier to follow now.

I went through the rest of the project and everything LGTM 👍

**Comment by naomi-lgbt** [comment]
Heya @moT01 is this still blocked? Or can we go ahead and get this in?

**Comment by raisedadead** [comment]
> can we go ahead and get this in?

I believe as soon as the helper solutions are ready on the guide we are good to go.

**Comment by ShaunSHamilton** [comment]
Are we at the point I should add the solutions to this PR?

**Review by Sboonny — DISMISSED** [review]

**Comment by naomi-lgbt** [comment]
I think this is actually blocked until https://github.com/freeCodeCamp/boilerplate-advancednode/pull/25 lands?

**Comment by moT01** [comment]
No, this needs to get in first, and once it hits production - we can merge that PR @naomi-lgbt.

**Review by moT01 — APPROVED** [review]

**Review by naomi-lgbt — APPROVED** [review]
Okay, this looks good. The tests all pass with Shaun's solutions (and the updated boilerplate), and I don't see any issues with the copy.

**Comment by Xavier-Pierre-dev** [comment]
Sorry I wasn't able to see all of the avancement on this PR before today. 

I also wanted to thanks @moT01, @ShaunSHamilton, @scissorsneedfoodtoo and all the contributors.

**Pull request merged**
